File: E2ETestResults/CoreSubset/Core/Logging/ConsoleLogAppender.cs
// ---------------------------------------------------------------------------------------------------------------------
// <copyright file="ConsoleLogAppender.cs" company="Tableau Software">
//   This file is the copyrighted property of Tableau Software and is protected by registered patents and other
//   applicable U.S. and international laws and regulations.
//   
//   Unlicensed use of the contents of this file is prohibited. Please refer to the NOTICES.txt file for further details.
// </copyright>
// ---------------------------------------------------------------------------------------------------------------------

namespace Tableau.JavaScript.Vql.Core
{
    using System;
    using System.Collections.Generic;
    using System.Diagnostics;
    using System.Html;
    using System.Runtime.CompilerServices;
    using Tableau.JavaScript.Vql.TypeDefs;

    /// <summary>
    /// An appender that writes to console.log.
    /// </summary>
    public class ConsoleLogAppender : BaseLogAppender
    {
        //// ===========================================================================================================
        //// Member Variables
        //// ===========================================================================================================

        private static ConsoleLogAppender globalAppender;

        private JsDictionary<string, object> levelMethods;

        //// ===========================================================================================================
        //// Constructors
        //// ===========================================================================================================

#if DEBUG

        static ConsoleLogAppender()
        {
            // by default we're going to enable console.log for Info+ in debug
            EnableLogging(delegate(Logger l, LoggerLevel ll)
            {
                return ll >= LoggerLevel.Info;
            });
        }

#endif

        private ConsoleLogAppender()
        {
        }

        //// ===========================================================================================================
        //// Methods
        //// ===========================================================================================================

        [AlternateSignature]
        public static extern void EnableLogging();

        /// <summary>
        /// Enables logging using this appender type.
        /// </summary>
        /// <param name="filter">The filter to apply to this appender or <c>null</c> to enable for all loggers</param>
        [Conditional("DEBUG")]
        public static void EnableLogging(Func<Logger, LoggerLevel, bool> filter)
        {
            if (Script.IsNullOrUndefined(globalAppender))
            {
                globalAppender = new ConsoleLogAppender();
                Logger.AddAppender(globalAppender);
            }

            globalAppender.AddFilter(ScriptEx.Value(filter, delegate { return true; }));
        }

        /// <summary>
        /// Disables logging using this appender type.
        /// </summary>
        [Conditional("DEBUG")]
        public static void DisableLogging()
        {
            if (Script.IsNullOrUndefined(globalAppender))
            {
                return;
            }

            Logger.RemoveAppender(globalAppender);
            globalAppender = null;
        }

        protected override void LogInternal(Logger source, LoggerLevel level, string message, object[] args)
        {
            if (Script.TypeOf(((dynamic)typeof(Window)).console) != "object")
            {
                return;
            }

            message = source.Name + ": " + message;

            JsArray<object> consoleArgs = new JsArray<object>();
            if (BrowserSupport.ConsoleLogFormating)
            {
                consoleArgs = consoleArgs.Concat(message).Concat(args);
            }
            else
            {
                // console.log doesn't do formatting so we need to
                consoleArgs = consoleArgs.Concat(this.FormatMessage(message, args));
            }

            try
            {
                ((dynamic)typeof(Function).Prototype).apply.call(this.GetConsoleMethod(level), ((dynamic)typeof(Window)).console, consoleArgs);
            }
            catch
            {
                // ignore
            }
        }

        private object GetConsoleMethod(LoggerLevel level)
        {
            dynamic console = TypeUtil.GetField<object>(Window.Self, "console");
            if (this.levelMethods == null)
            {
                this.levelMethods = new JsDictionary<string, object>();
                this.levelMethods[LoggerLevel.Debug.ToString()] = console.log;
                this.levelMethods[LoggerLevel.Error.ToString()] = console.error;
                this.levelMethods[LoggerLevel.Info.ToString()] = console.info;
                this.levelMethods[LoggerLevel.Warn.ToString()] = console.warn;
            }

            return this.levelMethods[level.ToString()] ?? console.log;
        }
    }
}

File: E2ETestResults/CoreSubset/Core/Logging/ConsoleLogAppender.ts
import { BaseLogAppender } from './BaseLogAppender';

import { BrowserSupport } from '../Utility/BrowserSupport';

import { Logger, LoggerLevel } from './Logger';

import 'mscorlib';

import { ScriptEx } from '../../CoreSlim/ScriptEx';

import { TypeUtil } from 'NativeJsTypeDefs';

/**
 * An appender that writes to console.log.
 */
export class ConsoleLogAppender extends BaseLogAppender {
  private static globalAppender: ConsoleLogAppender;

  private levelMethods: { [key: string]: any };

  /**
   */
  public static __ctor() {
    ConsoleLogAppender.enableLogging((l: Logger, ll: LoggerLevel) => {
      return ll >= LoggerLevel.Info;
    });
  }

  /**
   */
  private constructor() { }

  public static enableLogging(): void;

  /**
   * Enables logging using this appender type.
   * @param filter The filter to apply to this appender or `null` to enable for all loggers
   */
  public static enableLogging(filter?: (logger: Logger, loggerLevel: LoggerLevel) => boolean): void {
    if (ss.isNullOrUndefined(ConsoleLogAppender.globalAppender)) {
      ConsoleLogAppender.globalAppender = new ConsoleLogAppender();
      Logger.addAppender(ConsoleLogAppender.globalAppender);
    }
    ConsoleLogAppender.globalAppender.addFilter((filter) || (() => {
      return true;
    }));
  }

  /**
   * Disables logging using this appender type.
   */
  public static disableLogging(): void {
    if (ss.isNullOrUndefined(ConsoleLogAppender.globalAppender)) {
      return;
    }
    Logger.removeAppender(ConsoleLogAppender.globalAppender);
    ConsoleLogAppender.globalAppender = null;
  }

  protected logInternal(source: Logger, level: LoggerLevel, message: string, args: any[]): void {
    if (typeof (<any>window).console !== 'object') {
      return;
    }
    message = source.name + ': ' + message;
    let consoleArgs: any[] = [];
    if (BrowserSupport.consoleLogFormating) {
      consoleArgs = consoleArgs.concat(message).concat(args);
    } else {
      consoleArgs = consoleArgs.concat(this.formatMessage(message, args));
    }
    try {
      (<any>Function.prototype).apply.call(this.getConsoleMethod(level), (<any>window).console, consoleArgs);
    } catch { }
  }

  private getConsoleMethod(level: LoggerLevel): any {
    let console: any = window.self['console'];
    if (this.levelMethods === null) {
      this.levelMethods = {};
      this.levelMethods[LoggerLevel.Debug.toString()] = console.log;
      this.levelMethods[LoggerLevel.Error.toString()] = console.error;
      this.levelMethods[LoggerLevel.Info.toString()] = console.info;
      this.levelMethods[LoggerLevel.Warn.toString()] = console.warn;
    }
    return this.levelMethods[level.toString()] || console.log;
  }
}

// Call the static constructor
ConsoleLogAppender.__ctor();
